### DISCUSSION TITLE
MVI Architecture

### DISCUSSION THREAD

**Discussion by Yhdista** [adrielcafe/voyager#107] [Q&A]
Has anyone experienced with MVI + Voyager + ModeView. Problem is with ViewEffect/Action (usually navigation to somewhere, show toast etc.)
using ViewEffect.collectedAsState not working correctly 
and with using SideEffect or something then using Navigator is impossible as the Navigator is composable.
If somebody using MVI here, can I see your base classes?

**Comment by zakrodionov** [comment]
Just get navigator before SideEffect

```
val navigator = LocalNavigator.currentOrThrow

LaunchedEffect(Unit) {
    viewModel.eventFlow.collect {
        when (it) {
            NavigateToLoginFlow -> navigator.push(LoginFlow())
        }
    }
}

```

**Reply by Yhdista** [reply]
Thx.
As I needed general solution for base class (BaseScreen) so I did this:

```
val navigator = LocalNavigator.currentOrThrow

LaunchedEffect(Unit) {
    viewModel.effect.collect { effect ->
        onEffect(effect, navigator)
    }
}

open fun onEffect(effect: ViewEffect, navigator: Navigator) {}
```

**Comment by Alaksion** [comment]
Just retrieve your navigator and use it inside a launched effect. That's how I'm doing:
[Compose view](https://github.com/Alaksion/mvi_sample/blob/main/features/albums/create/src/main/kotlin/features/albums/create/presentation/steps/CreateAlbumScreen.kt)
[ViewModel](https://github.com/Alaksion/mvi_sample/blob/main/features/albums/create/src/main/kotlin/features/albums/create/presentation/ManageAlbumViewModel.kt)
[UiEvent](https://github.com/Alaksion/mvi_sample/blob/main/platform/uistate/src/main/java/platform/uistate/uievent/UiEventEffect.kt)

**Reply by Yhdista** [reply]
Your Screens (steps) are connected to the same ViewModel... what if every Screen has its own ViewModel... how would parent VM listens to children VM and vice versa?

**Reply by Alaksion** [reply]
What do you mean by child ViewModels?
